lib/elixir/lib/calendar.ex

```diff
@@ -789,28 +945,93 @@ defmodule Time do
   """
   @spec compare(Calendar.time, Calendar.time) :: :lt | :eq | :gt
   def compare(time1, time2) do
-    case {to_tuple(time1), to_tuple(time2)} do
+    {parts1, ppd1} = to_day_fraction(time1)
+    {parts2, ppd2} = to_day_fraction(time2)
+
+    case {parts1 * ppd2, parts2 * ppd1} do
       {first, second} when first > second -> :gt
       {first, second} when first < second -> :lt
       _ -> :eq
     end
   end
 
+  @doc """
+  Converts the Time struct to a different calendar.
+  """
+  @spec convert(Time.t, Calendar.calendar) :: Time.t
+  # No conversion required if already of the target calendar.
+  def convert(%Time{calendar: calendar} = time, calendar) do
+    {:ok, time}
+  end
+
+  def convert(%Time{} = time, calendar) do
+    result_time =
+      time
+      |> to_day_fraction
+      |> calendar.time_from_day_fraction
+    {:ok, result_time}
+  end
+
+  @spec convert!(Time.t, Calendar.calendar) :: Time.t
+  def convert!(time, calendar) do
+    case convert(time, calendar) do
+      {:ok, value} ->
+        value
+      {:error, reason} ->
+        raise ArgumentError, "cannot convert #{inspect time} to target calendar #{calendar}, reason: #{inspect reason}"
+    end
+  end
+
+  @doc """
+  Returns a day_fraction that represents the difference
+  between two Time Structs.
+  """
+  @spec diff(Time.t, Time.t) :: Calendar.rata_die
+  def diff(%Time{} = time1, %Time{} = time2) do
+    {parts1, ppd1} = to_day_fraction(time1)
+    {parts2, ppd2} = to_day_fraction(time2)
+
+    diff_ppd = ppd1 * ppd2
+    diff_parts = parts1 * ppd2 - parts2 * ppd1
+
+    # Keep integers in day fraction low.
+    gcd = gcd(diff_parts, diff_ppd)
+    diff_parts = div(diff_parts, gcd)
+    diff_ppd = div(diff_ppd, gcd)
+
+    {diff_parts, diff_ppd}
+  end
+
   ## Helpers
 
   defp to_tuple(%{hour: hour, minute: minute, second: second, microsecond: {microsecond, _precision}}) do
     {hour, minute, second, microsecond}
   end
 
+  defp to_day_fraction(%{hour: hour, minute: minute, second: second, microsecond: {_, _} = microsecond, calendar: calendar}) do
+    calendar.time_to_day_fraction(hour, minute, second, microsecond)
+  end
+
+  defp to_day_fraction(%{hour: hour, minute: minute, second: second, microsecond: microsecond, calendar: calendar}) do
+    calendar.time_to_day_fraction(hour, minute, second, {microsecond, 0})
+  end
+
+  @spec gcd(integer, integer) :: non_neg_integer
```

Comment:
We are not specing private functions anywhere in this module, so let's stay consistent here.